### PR TITLE
Debug BDS segfault #253

### PR DESCRIPTION
**Pull request by klin333**
Debugging and potential fix of BDS crashes from #253

Added a bds_debug function to send thousands of bds requests in rapid succession. This helps reliably and quickly reproduce the error. See also tinytests/test_segfault.R

My hunch is that unprotected SEXP in bds.cpp get destroyed by garbage collection. After changing SEXP to Rcpp::RObject throughout bds.cpp, I could no longer reproduce the error. However, I do not have sufficient theoretical knowledge of SEXP to know for sure.

### CODE REVIEW

**Comment by eddelbuettel** [comment]
Besides switching to `RObject`, we could also wrap `Shield<>` or `Protect<>` around the `SEXP` to have a more minimal change.  Otherwise concur.

**Comment by eddelbuettel** [comment]
Could you, if you have another minute or two, maybe try replacing the `SEXP foo` uses and assignments with `Rcpp::Shield<SEXP> foo` throughout?   That is done more commonly.

(`RObject` is / was good too. At one point it was the top of our object hierarchy but things got moved around. It is not that widely used / idiomatic Rcpp use these days.  Mostly stylistic though.  You put your needle on the issue.)

**Comment by klin333** [comment]
Might need some assistance with Shield, getting compiler errors everywhere, to the point where the only way for me to compile is to have no Shield and be back to where we began. 

Something like ```Rcpp::Shield<SEXP> foo;``` doesn't even compile with ```no matching function for call to 'Rcpp::Shield<SEXPREC*>::Shield()'```. I have very minimal knowledge about templates to figure this out... 

Also note that an earlier commit 7f3531d3 deliberately moved away from Shield.

**Comment by eddelbuettel** [comment]
I concur. I just gave it a spin on a checkout of your branch and that is a fight not worth fighting.  Switching to RObject is a net improvement, and you proposed a fine PR.  I say we go with it.

File: R/RcppExports.R [hunk 1]
```diff
@@ -17,6 +17,10 @@ bds_Impl <- function(con_, securities, field, options_, overrides_, verbose, ide
     .Call(`_Rblpapi_bds_Impl`, con_, securities, field, options_, overrides_, verbose, identity_)
 }
 
+bds_Impl_Debug <- function(con_, securities, field, options_, overrides_, repeats, verbose, identity_) {
```
Comment (eddelbuettel) [inline R/RcppExports.R:20]: As a PR to master, this is not needed.  Can you remove the function, and run `compileAttributes()` again to remove it here?

File: R/bds.R [hunk 2]
```diff
@@ -59,3 +59,14 @@ bds <- function(security, field, options=NULL,
     bds_Impl(con, security, field, options, overrides, verbose, identity)
 }
 
+
+bds_debug <- function(security, field, options=NULL,
```
Comment (eddelbuettel) [inline R/bds.R:63]: As above, I suggest to remove.  Keep the branch if you will or a copy but we do not this merged.

File: inst/tinytest/test_segfault.R [hunk 3]
```diff
@@ -0,0 +1,9 @@
+
+library(Rblpapi)
+
+blpConnect()
+
+res <- Rblpapi:::bds_debug("YCSW0303 Index", "par_curve", repeats = 10000)
+
+print("no segfault")
+print(length(res))
```
Comment (eddelbuettel) [inline inst/tinytest/test_segfault.R:9]: We need to either remove the test file (my suggestion) or condition it similar to the others.  Code can not simply access Blp, sadly.

File: src/RcppExports.cpp [hunk 4]
```diff
@@ -73,6 +73,24 @@ BEGIN_RCPP
     return rcpp_result_gen;
 END_RCPP
 }
+// bds_Impl_Debug
+Rcpp::List bds_Impl_Debug(SEXP con_, std::vector<std::string> securities, std::string field, SEXP options_, SEXP overrides_, int repeats, bool verbose, SEXP identity_);
```
Comment (eddelbuettel) [inline src/RcppExports.cpp:77]: As above, delete and regenerate so this goes away.

File: src/bds.cpp [hunk 5]
```diff
@@ -293,6 +293,66 @@ Rcpp::List bds_Impl(SEXP con_, std::vector<std::string> securities,
     return R_NilValue;
 }
 
+
+// debug: send many requests at once to raise chance of hitting segfault
+// [[Rcpp::export]]
+Rcpp::List bds_Impl_Debug(SEXP con_, std::vector<std::string> securities,
```
Comment (eddelbuettel) [inline src/bds.cpp:299]: As above, please remove and rerun `compileAttributes()`.

**Review by eddelbuettel — CHANGES_REQUESTED** [review]

**Comment by klin333** [comment]
Would it be easier if I make a new fork and just git cherry-pick that 1 commit which actually did the fix? Ie remove the bds_debug commit and remove the tinytest commit.

Eg https://github.com/klin333/Rblpapi/tree/bds_segfault

It looks like if I do that, i'll make a new pull request with the new branch which just has the single commit

**Comment by eddelbuettel** [comment]
Sure. 

(Random idea:  hard reset three commits back then just cherry pick the one commit in.  No idea if that works.)

Throwing it away and doing a clean PR is easiest. There are worse things than a closed PR.  (If you are redoing things please look at ChangeLog and pretty-please consider adding an entry with name and email.)

**Comment by klin333** [comment]
ok, redone as new pull request with just the fix, and ChangeLog entry.

**Comment by eddelbuettel** [comment]
So I guess we should / can close this one.

**Pull request closed without merging**